torch/lib/c10d/logger.cpp

```diff
@@ -85,6 +87,101 @@ void Logger::set_construction_logging_data(
   LogPyTorchDDPUsage(*ddp_logging_data_);
 }
 
+void Logger::calculate_avg_cpu_time(
+    int64_t& avg_time,
+    int64_t cpu_start_time,
+    int64_t cpu_end_time) {
+  long num_iters = reducer_->num_iterations_;
+  avg_time = ((cpu_end_time - cpu_start_time) + avg_time * (num_iters - 1)) /
+      num_iters;
+}
+
+#ifdef USE_CUDA
+void Logger::calculate_avg_gpu_time(
+    int64_t& avg_time,
+    cudaEvent_t gpu_start,
+    cudaEvent_t gpu_end) {
+  long num_iters = reducer_->num_iterations_;
+  float milliseconds = 0.0;
+  cudaEventElapsedTime(&milliseconds, gpu_start, gpu_end);
+  avg_time =
+      (int(milliseconds * 1000000) + avg_time * (num_iters - 1)) / num_iters;
+}
+#endif
+
+void Logger::set_runtime_stats() {
+  // set runtime stats after 1st iteration is complete.
+  if (reducer_->num_iterations_ == 0) {
+    return;
+  }
+  // Set ith iteration when the runtime stats are set.
+  ddp_logging_data_->iteration = reducer_->num_iterations_;
+  // If unused_parameters_ is not empty, calculate its sizes.
+  // unused_parameters_ is calculated in forward call of
+  // each iteration.
+  for (const auto& unused_index : reducer_->unused_parameters_) {
+    const auto& v = reducer_->replicas_[unused_index.replica_index]
+                                       [unused_index.variable_index];
+    ddp_logging_data_->unused_parameter_size += v.numel() * v.element_size();
+  }
+  // rebuilt_bucket_sizes will not change once buckets are rebuilt,
+  // so it only needs to set once during whole training loop.
+  if (reducer_->num_iterations_ == 2) {
+    ddp_logging_data_->has_rebuilt_buckets = reducer_->has_rebuilt_bucket_;
+    ddp_logging_data_->rebuilt_bucket_sizes = get_bucket_sizes();
+  }
+
+  // set_runtime_stats is called at beginning of forward call,
+  // when it is cheap to synchronize the cuda events of previous iteration,
+  // as it is guaranteed that all cuda operations are done in previous
+  // iteration.
+  if (reducer_->replicas_[0][0].is_cuda()) {
+#ifdef USE_CUDA
```

Comment:
This compile-time flag should always be true in this `if` statement right? Otherwise how could `is_cuda()` be true?

Reply:
Ah nevermind, one is compile time and the other is runtime